README.md
# go-piping-server
[Piping Server](https://github.com/nwtgck/piping-server) written in Go language (original: <https://github.com/nwtgck/piping-server>)

## Install for Ubuntu
```bash
wget https://github.com/nwtgck/go-piping-server/releases/download/v0.6.2/go-piping-server-0.6.2-linux-amd64.deb
sudo dpkg -i go-piping-server-0.6.2-linux-amd64.deb
```

## Install for macOS
```bash
brew install nwtgck/go-piping-server/go-piping-server
```

## Install for Windows
[Download](https://github.com/nwtgck/go-piping-server/releases/download/v0.6.2/go-piping-server-0.6.2-windows-amd64.zip)

Get more executables in the [releases](https://github.com/nwtgck/go-piping-server/releases).

## Docker

```bash
docker run -p 8181:8080 nwtgck/go-piping-server
```

## Server options

```
Infinitely transfer between any device over pure HTTP

Usage:
  go-piping-server [flags]

Flags:
      --crt-path string     Certification path
      --enable-http3        Enable HTTP/3 (experimental)
      --enable-https        Enable HTTPS
  -h, --help                help for go-piping-server
      --http-port uint16    HTTP port (default 8080)
      --https-port uint16   HTTPS port (default 8443)
      --key-path string     Private key path
      --version             show version
```


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)

## [Unreleased]

## [0.6.3] - 2023-09-13
### Changed
* Update dependencies

## [0.6.2] - 2023-08-09
### Changed
* Update dependencies

## [0.6.1] - 2023-04-09

## [0.6.0] - 2023-04-09
### Changed
* Update dependencies
* (Docker) golang:1.20

## [0.5.0] - 2023-03-08
### Changed
* Update dependencies
* (Docker) golang:1.19
* Smoother streaming
* Send message to sender

## [0.4.0] - 2022-01-15
### Added
* Support HEAD method for the reserved paths
* Add "Top page" link to /noscript page
* Log Piping Server version
* (Docker) Support linux/arm/v6, linux/arm/v7 and linux/arm64 Docker images

### Changed
* Update dependencies
* Add Go runtime version to --version output and version log

## [0.3.0] - 2021-11-20
### Added
* Make server logging better
* Support HTTP/2 Cleartext (h2c)
* Support X-Piping header passing arbitrary data from sender to receivers
* Create /noscript Web UI for transferring a file without JavaScript
* Support multipart upload

### Changed
* Use requested protocol in /help
* Add `X-Robots-Tag: "none"` header to receiver's response
* Reject POST and PUT with Content-Range for now to detect resumable upload in the future
* Respond 405 Method Not Allowed when method is not supported
* Reject Service Worker registration request

### Fixed
* Not allow receiver do GET a path while transferring
* Add `Access-Control-Allow-Origin: *` to sender's response

## [0.2.0] - 2021-09-18
### Added
* Support preflight request

## 0.1.0 - 2021-05-18
### Added
* Initial release

[Unreleased]: https://github.com/nwtgck/go-piping-server/compare/v0.6.3...HEAD
[0.6.3]: https://github.com/nwtgck/go-piping-server/compare/v0.6.2...v0.6.3
[0.6.2]: https://github.com/nwtgck/go-piping-server/compare/v0.6.1...v0.6.2
[0.6.1]: https://github.com/nwtgck/go-piping-server/compare/v0.6.0...v0.6.1
[0.6.0]: https://github.com/nwtgck/go-piping-server/compare/v0.5.0...v0.6.0
[0.5.0]: https://github.com/nwtgck/go-piping-server/compare/v0.4.0...v0.5.0
[0.4.0]: https://github.com/nwtgck/go-piping-server/compare/v0.3.0...v0.4.0
[0.3.0]: https://github.com/nwtgck/go-piping-server/compare/v0.2.0...v0.3.0
[0.2.0]: https://github.com/nwtgck/go-piping-server/compare/v0.1.0...v0.2.0


piping_server_test.go
package piping_server

import (
	"fmt"
	"github.com/nwtgck/go-piping-server/version"
	"golang.org/x/net/context"
	"gotest.tools/v3/assert"
	"io"
	"log"
	"net"
	"net/http"
	"strconv"
	"strings"
	"testing"
)

// serve serves Piping Server on available port
func serve(t *testing.T) (*http.Server, string) {
	logger := log.New(io.Discard, "", log.LstdFlags|log.Lmicroseconds)
	pipingServer := NewServer(logger)
	ln, err := net.Listen("tcp", "localhost:0")
	if err != nil {
		t.Fatal(err)
	}
	server := &http.Server{Handler: http.HandlerFunc(pipingServer.Handler)}
	go func() {
		err := server.Serve(ln)
		if err == http.ErrServerClosed {
			return
		}
		if err != nil {
			t.Error(err)
			return
		}
	}()
	return server, "http://" + ln.Addr().String()
}

func readerToString(t *testing.T, r io.Reader) string {
	stringBuilder := new(strings.Builder)
	_, err := io.Copy(stringBuilder, r)
	if err != nil {
		t.Fatal(err)
	}
	return stringBuilder.String()
}

func TestIndexPage(t *testing.T) {
	server, url := serve(t)
	defer server.Shutdown(context.Background())

	res, err := http.Get(url)
	if err != nil {
		t.Fatal(t)
	}
	assert.Equal(t, res.StatusCode, 200)
	assert.Equal(t, res.Header.Get("Content-Type"), "text/html")
	body := readerToString(t, res.Body)
	assert.Assert(t, strings.Contains(body, "Piping"))
}

func TestNoscriptPage(t *testing.T) {
	server, url := serve(t)
	defer server.Shutdown(context.Background())

	res, err := http.Get(url + "/noscript?path=mypath")
	if err != nil {
		t.Fatal(t)
	}
	assert.Equal(t, res.StatusCode, 200)
	assert.Equal(t, res.Header.Get("Content-Type"), "text/html")
	body := readerToString(t, res.Body)
	assert.Assert(t, strings.Contains(body, `action="mypath"`))
}

func TestVersionPage(t *testing.T) {
	server, url := serve(t)
	defer server.Shutdown(context.Background())

	res, err := http.Get(url + "/version")
	if err != nil {
		t.Fatal(t)
	}
	assert.Equal(t, res.StatusCode, 200)
	assert.Equal(t, res.Header.Get("Content-Type"), "text/plain")
	assert.Equal(t, res.Header.Get("Access-Control-Allow-Origin"), "*")
	body := readerToString(t, res.Body)
	assert.Equal(t, body, fmt.Sprintf("%s in Go\n", version.Version))
}

func TestHelpPage(t *testing.T) {
	server, url := serve(t)
	defer server.Shutdown(context.Background())

	res, err := http.Get(url + "/help")
	if err != nil {
		t.Fatal(t)
	}
	assert.Equal(t, res.StatusCode, 200)
	assert.Equal(t, res.Header.Get("Content-Type"), "text/plain")
	assert.Equal(t, res.Header.Get("Access-Control-Allow-Origin"), "*")
}

func TestFavicon(t *testing.T) {
	server, url := serve(t)
	defer server.Shutdown(context.Background())

	res, err := http.Get(url + "/favicon.ico")
	if err != nil {
		t.Fatal(t)
	}
	assert.Equal(t, res.StatusCode, 204)
	bytes, err := io.ReadAll(res.Body)
	if err != nil {
		t.Fatal(t)
	}
	assert.Equal(t, len(bytes), 0)
}

func TestRobotsTxt(t *testing.T) {
	server, url := serve(t)
	defer server.Shutdown(context.Background())

	res, err := http.Get(url + "/robots.txt")
	if err != nil {
		t.Fatal(t)
	}
	assert.Equal(t, res.StatusCode, 404)
	bytes, err := io.ReadAll(res.Body)
	if err != nil {
		t.Fatal(t)
	}
	assert.Equal(t, len(bytes), 0)
}

func TestHeadMethodInReservedPaths(t *testing.T) {
	server, url := serve(t)
	defer server.Shutdown(context.Background())

	for _, path := range reservedPaths {
		headRes, err := http.Head(url + path)
		if err != nil {
			t.Fatal(t)
		}
		getRes, err := http.Get(url + path)
		if err != nil {
			t.Fatal(t)
		}
		headRes.Header.Del("Date")
		getRes.Header.Del("Date")
		assert.Equal(t, headRes.StatusCode, getRes.StatusCode)
		assert.DeepEqual(t, headRes.Header, getRes.Header)
	}
}

func TestPreflightRequest(t *testing.T) {
	server, url := serve(t)
	defer server.Shutdown(context.Background())

	req, err := http.NewRequest("OPTIONS", url+"/mypath", nil)
	if err != nil {
		t.Fatal(t)
	}
	res, err := http.DefaultClient.Do(req)
	if err != nil {
		t.Fatal(t)
	}
	assert.Equal(t, res.StatusCode, 200)
	assert.Equal(t, res.Header.Get("Access-Control-Allow-Origin"), "*")
	assert.Equal(t, res.Header.Get("Access-Control-Allow-Methods"), "GET, HEAD, POST, PUT, OPTIONS")
	assert.Equal(t, strings.ToLower(res.Header.Get("Access-Control-Allow-Headers")), "content-type, content-disposition, x-piping")
	assert.Equal(t, res.Header.Get("Access-Control-Max-Age"), "86400")
}

func TestRejectServiceWorkerRegistration(t *testing.T) {
	server, url := serve(t)
	defer server.Shutdown(context.Background())

	req, err := http.NewRequest("GET", url+"/mysw.js", nil)
	if err != nil {
		t.Fatal(t)
	}
	req.Header.Set("Service-Worker", "script")
	res, err := http.DefaultClient.Do(req)
	if err != nil {
		t.Fatal(t)
	}
	assert.Equal(t, res.StatusCode, 400)
	assert.Equal(t, res.Header.Get("Access-Control-Allow-Origin"), "*")
}

func TestRejectRangeAccessForNow(t *testing.T) {
	server, url := serve(t)
	defer server.Shutdown(context.Background())

	for _, method := range []string{"POST", "PUT"} {
		req, err := http.NewRequest(method, url+"/mypath", strings.NewReader("hello"))
		if err != nil {
			t.Fatal(t)
		}
		req.Header.Set("Content-Range", "bytes 2-6/100")
		res, err := http.DefaultClient.Do(req)
		if err != nil {
			t.Fatal(t)
		}
		assert.Equal(t, res.StatusCode, 400)
		assert.Equal(t, res.Header.Get("Access-Control-Allow-Origin"), "*")
	}
}

func TestTransferSenderReceiver(t *testing.T) {
	server, url := serve(t)
	defer server.Shutdown(context.Background())

	sendBodyStr := "this is a content"
	senderReq, err := http.NewRequest("POST", url+"/mypath", strings.NewReader(sendBodyStr))
	if err != nil {
		t.Fatal(t)
	}
	senderReq.Header.Set("Content-Type", "text/plain")
	senderResCh := make(chan *http.Response, 1)
	{
		res, err := http.DefaultClient.Do(senderReq)
		if err != nil {
			t.Error(t)
			return
		}
		senderResCh <- res
	}
	receiverReq, err := http.NewRequest("GET", url+"/mypath", nil)
	if err != nil {
		t.Fatal(t)
	}
	receiverRes, err := http.DefaultClient.Do(receiverReq)
	senderRes := <-senderResCh
	assert.Equal(t, senderRes.StatusCode, 200)
	assert.Equal(t, senderRes.Header.Get("Access-Control-Allow-Origin"), "*")

	assert.Equal(t, receiverRes.StatusCode, 200)
	assert.Equal(t, receiverRes.Header.Get("Content-Type"), "text/plain")
	assert.Equal(t, receiverRes.Header.Get("Content-Length"), strconv.Itoa(len(sendBodyStr)))
	assert.Assert(t, len(receiverRes.Header.Values("Content-Disposition")) == 0)
	assert.Equal(t, receiverRes.Header.Get("Access-Control-Allow-Origin"), "*")
	assert.Equal(t, receiverRes.Header.Get("X-Robots-Tag"), "none")
	assert.Assert(t, len(receiverRes.Header.Values("Access-Control-Expose-Headers")) == 0)
}

func TestTransferWithoutSniff(t *testing.T) {
	server, url := serve(t)
	defer server.Shutdown(context.Background())

	sendBodyStr := "this is a content"
	senderReq, err := http.NewRequest("POST", url+"/mypath", strings.NewReader(sendBodyStr))
	if err != nil {
		t.Fatal(t)
	}
	// NOTE: No content-type is specified in senderReq
	senderResCh := make(chan *http.Response)
	go func() {
		res, err := http.DefaultClient.Do(senderReq)
		if err != nil {
			t.Error(t)
			return
		}
		senderResCh <- res
	}()
	receiverReq, err := http.NewRequest("GET", url+"/mypath", nil)
	if err != nil {
		t.Fatal(t)
	}
	receiverRes, err := http.DefaultClient.Do(receiverReq)
	senderRes := <-senderResCh
	assert.Equal(t, senderRes.StatusCode, 200)

	assert.Equal(t, receiverRes.StatusCode, 200)
	assert.Assert(t, len(receiverRes.Header.Values("Content-Type")) == 0)
}

func TestTransferReceiverSender(t *testing.T) {
	server, url := serve(t)
	defer server.Shutdown(context.Background())

	receiverReq, err := http.NewRequest("GET", url+"/mypath", nil)
	if err != nil {
		t.Fatal(t)
	}
	receiverResCh := make(chan *http.Response)
	go func() {
		res, err := http.DefaultClient.Do(receiverReq)
		if err != nil {
			t.Error(t)
			return
		}
		receiverResCh <- res
	}()

	sendBodyStr := "this is a content"
	senderReq, err := http.NewRequest("POST", url+"/mypath", strings.NewReader(sendBodyStr))
	if err != nil {
		t.Fatal(t)
	}
	senderReq.Header.Set("Content-Type", "text/plain")
	senderRes, err := http.DefaultClient.Do(senderReq)
	if err != nil {
		t.Error(t)
		return
	}

	assert.Equal(t, senderRes.StatusCode, 200)
	assert.Equal(t, senderRes.Header.Get("Access-Control-Allow-Origin"), "*")

	receiverRes := <-receiverResCh
	assert.Equal(t, receiverRes.StatusCode, 200)
	assert.Equal(t, receiverRes.Header.Get("Content-Type"), "text/plain")
	assert.Equal(t, receiverRes.Header.Get("Content-Length"), strconv.Itoa(len(sendBodyStr)))
	assert.Assert(t, len(receiverRes.Header.Values("Content-Disposition")) == 0)
	assert.Equal(t, receiverRes.Header.Get("Access-Control-Allow-Origin"), "*")
	assert.Equal(t, receiverRes.Header.Get("X-Robots-Tag"), "none")
	assert.Assert(t, len(receiverRes.Header.Values("Access-Control-Expose-Headers")) == 0)
}

func TestTransferXPipingHeaders(t *testing.T) {
	server, url := serve(t)
	defer server.Shutdown(context.Background())

	sendBodyStr := "this is a content"
	senderReq, err := http.NewRequest("POST", url+"/mypath", strings.NewReader(sendBodyStr))
	if err != nil {
		t.Fatal(t)
	}
	senderReq.Header.Set("Content-Type", "text/plain")
	senderReq.Header.Add("X-Piping", "mymetadata1")
	senderReq.Header.Add("X-Piping", "mymetadata2")
	senderReq.Header.Add("X-Piping", "mymetadata3")
	senderResCh := make(chan *http.Response)
	go func() {
		res, err := http.DefaultClient.Do(senderReq)
		if err != nil {
			t.Error(t)
			return
		}
		senderResCh <- res
	}()
	receiverReq, err := http.NewRequest("GET", url+"/mypath", nil)
	if err != nil {
		t.Fatal(t)
	}
	receiverRes, err := http.DefaultClient.Do(receiverReq)
	senderRes := <-senderResCh
	assert.Equal(t, senderRes.StatusCode, 200)
	assert.Equal(t, senderRes.Header.Get("Access-Control-Allow-Origin"), "*")

	assert.Equal(t, receiverRes.StatusCode, 200)
	assert.Equal(t, receiverRes.Header.Get("Content-Type"), "text/plain")
	assert.Equal(t, receiverRes.Header.Get("Content-Length"), strconv.Itoa(len(sendBodyStr)))
	assert.Assert(t, len(receiverRes.Header.Values("Content-Disposition")) == 0)
	assert.Equal(t, receiverRes.Header.Get("Access-Control-Allow-Origin"), "*")
	assert.Equal(t, receiverRes.Header.Get("X-Robots-Tag"), "none")
	assert.Equal(t, receiverRes.Header.Get("Access-Control-Expose-Headers"), "X-Piping")
	assert.DeepEqual(t, receiverRes.Header.Values("X-Piping"), []string{"mymetadata1", "mymetadata2", "mymetadata3"})
}


cmd/root.go
package cmd

import (
	"errors"
	"fmt"
	"github.com/nwtgck/go-piping-server"
	"github.com/nwtgck/go-piping-server/version"
	"github.com/quic-go/quic-go/http3"
	"github.com/spf13/cobra"
	"golang.org/x/net/http2"
	"golang.org/x/net/http2/h2c"
	"log"
	"net/http"
	"os"
	"runtime"
)

var showsVersion bool
var httpPort uint16
var enableHttps bool
var httpsPort uint16
var keyPath string
var crtPath string
var enableHttp3 bool

func init() {
	cobra.OnInitialize()
	RootCmd.PersistentFlags().BoolVarP(&showsVersion, "version", "", false, "show version")
	RootCmd.PersistentFlags().Uint16VarP(&httpPort, "http-port", "", 8080, "HTTP port")
	RootCmd.PersistentFlags().BoolVarP(&enableHttps, "enable-https", "", false, "Enable HTTPS")
	RootCmd.PersistentFlags().Uint16VarP(&httpsPort, "https-port", "", 8443, "HTTPS port")
	RootCmd.PersistentFlags().StringVarP(&keyPath, "key-path", "", "", "Private key path")
	RootCmd.PersistentFlags().StringVarP(&crtPath, "crt-path", "", "", "Certification path")
	RootCmd.PersistentFlags().BoolVarP(&enableHttp3, "enable-http3", "", false, "Enable HTTP/3 (experimental)")
}

var RootCmd = &cobra.Command{
	Use:          os.Args[0],
	Short:        "piping-server",
	Long:         "Infinitely transfer between any device over pure HTTP",
	SilenceUsage: true,
	RunE: func(cmd *cobra.Command, args []string) error {
		if showsVersion {
			fmt.Printf("%s (%s)\n", version.Version, runtime.Version())
			return nil
		}
		logger := log.New(os.Stderr, "", log.LstdFlags|log.Lmicroseconds)
		logger.Printf("Piping Server %s (%s)", version.Version, runtime.Version())
		pipingServer := piping_server.NewServer(logger)
		errCh := make(chan error)
		if enableHttps || enableHttp3 {
			if keyPath == "" {
				return errors.New("--key-path should be specified")
			}
			if crtPath == "" {
				return errors.New("--crt-path should be specified")
			}
			go func() {
				logger.Printf("Listening HTTPS on %d...\n", httpsPort)
				errCh <- http.ListenAndServeTLS(fmt.Sprintf(":%d", httpsPort), crtPath, keyPath, http.HandlerFunc(pipingServer.Handler))
			}()
			if enableHttp3 {
				go func() {
					logger.Printf("Listening HTTP/3 on %d...\n", httpsPort)
					errCh <- http3.ListenAndServeQUIC(fmt.Sprintf(":%d", httpsPort), crtPath, keyPath, http.HandlerFunc(pipingServer.Handler))
				}()
			}
		}
		go func() {
			server := &http.Server{
				Addr:    fmt.Sprintf(":%d", httpPort),
				Handler: h2c.NewHandler(http.HandlerFunc(pipingServer.Handler), &http2.Server{}),
			}
			logger.Printf("Listening HTTP on %d...\n", httpPort)
			errCh <- server.ListenAndServe()
		}()
		return <-errCh
	},
}


piping_server.go
package piping_server

import (
	"fmt"
	"github.com/nwtgck/go-piping-server/version"
	"io"
	"log"
	"mime"
	"mime/multipart"
	"net/http"
	"net/textproto"
	"strconv"
	"sync"
	"sync/atomic"
)

const (
	reservedPathIndex      = "/"
	reservedPathNoScript   = "/noscript"
	reservedPathVersion    = "/version"
	reservedPathHelp       = "/help"
	reservedPathFaviconIco = "/favicon.ico"
	reservedPathRobotsTxt  = "/robots.txt"
)

var reservedPaths = [...]string{
	reservedPathIndex,
	reservedPathVersion,
	reservedPathHelp,
	reservedPathFaviconIco,
	reservedPathRobotsTxt,
}

const noscriptPathQueryParameterName = "path"

type pipe struct {
	receiverResWriterCh chan http.ResponseWriter
	sendFinishedCh      chan struct{}
	isSenderConnected   uint32 // NOTE: for atomic operation
	isTransferring      uint32 // NOTE: for atomic operation
}

type PipingServer struct {
	pathToPipe map[string]*pipe
	mutex      *sync.Mutex
	logger     *log.Logger
}

func isReservedPath(path string) bool {
	for _, p := range reservedPaths {
		if p == path {
			return true
		}
	}
	return false
}

func NewServer(logger *log.Logger) *PipingServer {
	return &PipingServer{
		pathToPipe: map[string]*pipe{},
		mutex:      new(sync.Mutex),
		logger:     logger,
	}
}

func (s *PipingServer) getPipe(path string) *pipe {
	// Set pipe if not found on the path
	s.mutex.Lock()
	defer s.mutex.Unlock()
	if _, ok := s.pathToPipe[path]; !ok {
		pi := &pipe{
			receiverResWriterCh: make(chan http.ResponseWriter, 1),
			sendFinishedCh:      make(chan struct{}),
			isSenderConnected:   0,
		}
		s.pathToPipe[path] = pi
		return pi
	}
	return s.pathToPipe[path]
}

func transferHeaderIfExists(w http.ResponseWriter, reqHeader textproto.MIMEHeader, header string) {
	values := reqHeader.Values(header)
	if len(values) == 1 {
		w.Header().Add(header, values[0])
	}
}

func getTransferHeaderAndBody(req *http.Request) (textproto.MIMEHeader, io.ReadCloser) {
	mediaType, params, mediaTypeParseErr := mime.ParseMediaType(req.Header.Get("Content-Type"))
	// If multipart upload
	if mediaTypeParseErr == nil && mediaType == "multipart/form-data" {
		multipartReader := multipart.NewReader(req.Body, params["boundary"])
		part, err := multipartReader.NextPart()
		if err != nil {
			// Return normal if multipart error
			return textproto.MIMEHeader(req.Header), req.Body
		}
		return part.Header, part
	}
	return textproto.MIMEHeader(req.Header), req.Body
}

func (s *PipingServer) Handler(resWriter http.ResponseWriter, req *http.Request) {
	s.logger.Printf("%s %s %s", req.Method, req.URL, req.Proto)
	path := req.URL.Path

	if req.Method == "GET" || req.Method == "HEAD" {
		switch path {
		case reservedPathIndex:
			indexPageBytes := []byte(indexPage)
			resWriter.Header().Set("Content-Type", "text/html")
			resWriter.Header().Set("Content-Length", strconv.Itoa(len(indexPageBytes)))
			resWriter.Header().Set("Access-Control-Allow-Origin", "*")
			resWriter.Write(indexPageBytes)
			return
		case reservedPathNoScript:
			noScriptHtmlBytes := []byte(noScriptHtml(req.URL.Query().Get(noscriptPathQueryParameterName)))
			resWriter.Header().Set("Content-Type", "text/html")
			resWriter.Header().Set("Content-Length", strconv.Itoa(len(noScriptHtmlBytes)))
			resWriter.Header().Set("Access-Control-Allow-Origin", "*")
			resWriter.Write(noScriptHtmlBytes)
			return
		case reservedPathVersion:
			versionBytes := []byte(fmt.Sprintf("%s in Go\n", version.Version))
			resWriter.Header().Set("Content-Type", "text/plain")
			resWriter.Header().Set("Content-Length", strconv.Itoa(len(versionBytes)))
			resWriter.Header().Set("Access-Control-Allow-Origin", "*")
			resWriter.Write(versionBytes)
			return
		case reservedPathHelp:
			protocol := "http"
			if req.TLS != nil {
				protocol = "https"
			}
			url := fmt.Sprintf(protocol+"://%s", req.Host)
			helpPageBytes := []byte(helpPage(url))
			resWriter.Header().Set("Content-Type", "text/plain")
			resWriter.Header().Set("Content-Length", strconv.Itoa(len(helpPageBytes)))
			resWriter.Header().Set("Access-Control-Allow-Origin", "*")
			resWriter.Write(helpPageBytes)
			return
		case reservedPathFaviconIco:
			resWriter.Header().Set("Content-Length", "0")
			resWriter.WriteHeader(204)
			return
		case reservedPathRobotsTxt:
			resWriter.Header().Set("Content-Length", "0")
			resWriter.WriteHeader(404)
			return
		}
	}

	// TODO: should close if either sender or receiver closes
	switch req.Method {
	case "GET":
		// If the receiver requests Service Worker registration
		// (from: https://speakerdeck.com/masatokinugawa/pwa-study-sw?slide=32)
		if req.Header.Get("Service-Worker") == "script" {
			resWriter.Header().Set("Access-Control-Allow-Origin", "*")
			resWriter.WriteHeader(400)
			resWriter.Write([]byte("[ERROR] Service Worker registration is rejected.\n"))
			return
		}
		pi := s.getPipe(path)
		// If already get the path or transferring
		if len(pi.receiverResWriterCh) != 0 || atomic.LoadUint32(&pi.isTransferring) == 1 {
			resWriter.Header().Set("Access-Control-Allow-Origin", "*")
			resWriter.WriteHeader(400)
			resWriter.Write([]byte("[ERROR] The number of receivers has reached limits.\n"))
			return
		}
		pi.receiverResWriterCh <- resWriter
		// Wait for finish
		<-pi.sendFinishedCh
	case "POST", "PUT":
		// If reserved path
		if isReservedPath(path) {
			resWriter.Header().Set("Access-Control-Allow-Origin", "*")
			resWriter.WriteHeader(400)
			resWriter.Write([]byte(fmt.Sprintf("[ERROR] Cannot send to the reserved path '%s'. (e.g. '/mypath123')\n", path)))
			return
		}
		// Notify that Content-Range is not supported
		// In the future, resumable upload using Content-Range might be supported
		// ref: https://github.com/httpwg/http-core/pull/653
		if len(req.Header.Values("Content-Range")) != 0 {
			resWriter.Header().Set("Access-Control-Allow-Origin", "*")
			resWriter.WriteHeader(400)
			resWriter.Write([]byte(fmt.Sprintf("[ERROR] Content-Range is not supported for now in %s\n", req.Method)))
			return
		}
		pi := s.getPipe(path)
		// If a sender is already connected
		if !atomic.CompareAndSwapUint32(&pi.isSenderConnected, 0, 1) {
			resWriter.Header().Set("Access-Control-Allow-Origin", "*")
			resWriter.WriteHeader(400)
			reqContentLength := req.ContentLength
			// NOTE: `req.ContentLength = 0` is a workaround for full duplex
			// Replace with https://github.com/golang/go/blob/457fd1d52d17fc8e73d4890150eadab3128de64d/src/net/http/responsecontroller.go#L119-L141 in the future
			req.ContentLength = 0
			if f, ok := resWriter.(http.Flusher); ok {
				f.Flush()
			}
			req.ContentLength = reqContentLength
			resWriter.Write([]byte(fmt.Sprintf("[ERROR] Another sender has been connected on '%s'.\n", path)))
			return
		}

		contentLength := req.ContentLength
		// NOTE: `req.ContentLength = 0` is a workaround for full duplex
		// Replace with https://github.com/golang/go/blob/457fd1d52d17fc8e73d4890150eadab3128de64d/src/net/http/responsecontroller.go#L119-L141 in the future
		req.ContentLength = 0
		resWriter.Header().Set("Access-Control-Allow-Origin", "*")
		resWriter.WriteHeader(200)
		if f, ok := resWriter.(http.Flusher); ok {
			f.Flush()
		}
		req.ContentLength = contentLength

		resWriteFlusher := NewWriteFlusherIfPossible(resWriter)
		if _, err := resWriteFlusher.Write([]byte("[INFO] Waiting for 1 receiver(s)...\n")); err != nil {
			return
		}
		receiverResWriter := <-pi.receiverResWriterCh
		if _, err := resWriteFlusher.Write([]byte("[INFO] A receiver was connected.\n")); err != nil {
			return
		}
		if _, err := resWriteFlusher.Write([]byte("[INFO] Start sending to 1 receiver(s)!\n")); err != nil {
			return
		}
		atomic.StoreUint32(&pi.isTransferring, 1)
		transferHeader, transferBody := getTransferHeaderAndBody(req)
		receiverResWriter.Header()["Content-Type"] = nil // not to sniff
		transferHeaderIfExists(receiverResWriter, transferHeader, "Content-Type")
		transferHeaderIfExists(receiverResWriter, transferHeader, "Content-Length")
		transferHeaderIfExists(receiverResWriter, transferHeader, "Content-Disposition")
		xPipingValues := req.Header.Values("X-Piping")
		if len(xPipingValues) != 0 {
			receiverResWriter.Header()["X-Piping"] = xPipingValues
		}
		receiverResWriter.Header().Set("Access-Control-Allow-Origin", "*")
		if len(xPipingValues) != 0 {
			receiverResWriter.Header().Set("Access-Control-Expose-Headers", "X-Piping")
		}
		receiverResWriter.Header().Set("X-Robots-Tag", "none")
		receiverResWriteFlusher := NewWriteFlusherIfPossible(receiverResWriter)
		if _, err := io.Copy(receiverResWriteFlusher, transferBody); err != nil {
			return
		}
		if _, err := resWriteFlusher.Write([]byte("[INFO] Sent successfully!\n")); err != nil {
			return
		}
		pi.sendFinishedCh <- struct{}{}
		delete(s.pathToPipe, path)
	case "OPTIONS":
		resWriter.Header().Set("Access-Control-Allow-Origin", "*")
		resWriter.Header().Set("Access-Control-Allow-Methods", "GET, HEAD, POST, PUT, OPTIONS")
		resWriter.Header().Set("Access-Control-Allow-Headers", "Content-Type, Content-Disposition, X-Piping")
		resWriter.Header().Set("Access-Control-Max-Age", "86400")
		resWriter.Header().Set("Content-Length", "0")
		resWriter.WriteHeader(200)
		return
	default:
		resWriter.WriteHeader(405)
		resWriter.Header().Set("Access-Control-Allow-Origin", "*")
		resWriter.Write([]byte(fmt.Sprintf("[ERROR] Unsupported method: %s.\n", req.Method)))
		return
	}
	s.logger.Printf("Transferring %s has finished in %s method.\n", req.URL.Path, req.Method)
}

type WriteFlusher struct {
	writer  io.Writer
	flusher http.Flusher
}

func NewWriteFlusherIfPossible(w http.ResponseWriter) io.Writer {
	if f, ok := w.(http.Flusher); ok {
		return &WriteFlusher{writer: w, flusher: f}
	}
	return w
}

func (f *WriteFlusher) Write(p []byte) (int, error) {
	n, err := f.writer.Write(p)
	if err != nil {
		return n, err
	}
	f.flusher.Flush()
	return n, err
}
